src/main/scala/com.snowplowanalytics.s3/loader/NsqSourceExecutor.scala

```diff
@@ -0,0 +1,154 @@
+
+// Joda-Time
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.DateTimeFormat
+
+// Logging
+import org.slf4j.LoggerFactory
+
+// This project
+import sinks._
+import serializers._
+import model._
+
+/**
+ * Executor for NSQSource
+ *
+ * @param config S3Loader configuration
+ * @param provider AWSCredentialsProvider
+ * @param badSink Configured BadSink
+ * @param serializer Serializer instance 
+ * @param maxConnectionTime Max time for trying to connect S3 instance 
+ */
+class NsqSourceExecutor(
+  config: S3LoaderConfig,
+  provider: AWSCredentialsProvider,
+  badSink: ISink,
+  serializer: ISerializer, 
+  maxConnectionTime: Long,
+  tracker: Option[Tracker]
+) extends Runnable {
+
+  lazy val log = LoggerFactory.getLogger(getClass())
+
+  //nsq messages will be buffered in msgBuffer until buffer size become equal to nsqBufferSize
+  val msgBuffer = new ListBuffer[EmitterInput]()
+
+  val s3Emitter = new S3Emitter(config.s3, provider, badSink, maxConnectionTime, tracker)
+  private val TimeFormat = DateTimeFormat.forPattern("HH:mm:ss.SSS").withZone(DateTimeZone.UTC)
+  private val DateFormat = DateTimeFormat.forPattern("yyyy-MM-dd").withZone(DateTimeZone.UTC)
+
+  private def getBaseFilename(startTime: Long, endTime: Long): String = {
+    def abs (e: Int) = if (e > 0) e else -e
+    val currentTimeObject = new DateTime(System.currentTimeMillis())
+    val startTimeObject = new DateTime(startTime)
+    val endTimeObject = new DateTime(endTime)
+    val rand = Random
+    val randNum = rand.nextInt
+
+    DateFormat.print(currentTimeObject) + "-" + 
+      TimeFormat.print(startTimeObject) + "-" +
+      TimeFormat.print(endTimeObject)   + "-" + abs(randNum)
+  }
+  
+  override def run: Unit = {
+
+    val nsqCallback = new  NSQMessageCallback {
+      //start time of filling the buffer
+      var bufferStartTime = System.currentTimeMillis()
+      val nsqBufferSize = config.streams.buffer.recordLimit
+
+      override def message(msg: NSQMessage): Unit = {
+        val validMsg = msg.getMessage.success
+        msgBuffer.synchronized {
+          msgBuffer += validMsg
+          msg.finished()
+          if (msgBuffer.size >= nsqBufferSize) {
+            //finish time of filling the buffer
+            val bufferEndTime = System.currentTimeMillis()
+            val baseFilename = getBaseFilename(bufferStartTime, bufferEndTime)
+            val serializationResults = serializer.serialize(msgBuffer.toList, baseFilename)
+            val (successes, failures) = serializationResults.results.partition(_.isSuccess)
+
+            if (successes.size > 0) {
+              serializationResults.namedStreams.foreach { 
+                val connectionAttemptStartTime = System.currentTimeMillis()
+                s3Emitter.attemptEmit(_, connectionAttemptStartTime) match {
+                  case false => log.error(s"Error while sending to S3")
+                  case true => 
```

Comment:
could we add logging messages here and above similar to:

- https://github.com/snowplow/snowplow-s3-loader/blob/master/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/S3Emitter.scala#L123
- https://github.com/snowplow/snowplow-s3-loader/blob/master/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/S3Emitter.scala#L130

and one here saying that we successfully sent x messages to s3?